### PR TITLE
Kindness is a virtue: changing precise AST kinds to support diffing, comments, partial passes and more!

### PR DESCRIPTION
**Pull request by aymannadeem**
This PR introduces refactors that simplify readability, and will also make it easier to transition ASTs to kind `(* -> *) -> (* -> *)`. This entails:

- [x] generate `f :: * -> *` parameters for the record types, in TH, and add them to the generated datatype
- [x] fix the `Traversable`, `Foldable`, and `Functor` instances to account for the new kindedness
- [x] ensure that the children of each record types are wrapped in an `f` (except for the `ann` field)
- [x] define some Error functor isomorphic to `Either String`
- [x] manually define applicative instance for Error type
- [ ] add unit tests
- [ ] use Error type in `Unmarshal`

## Why is this important?

This addresses #480, #481, #482. 

TL;DR: it's a byproduct of having strongly-typed trees, where we can't take advantage of historic polymorphic-ness of our nodes and use `Error` types everywhere. 

@patrickt explained this to me using the following example. Consider a data type like this which looks like the ALC syntax:

```Haskell
data Plus a = Plus { ann :: a, left :: Expression a, right :: Expression a }
```

That type is polymorphic in its annotation type `a`, but its children are monomorphic. They are always `Expression` values. And if we get a parse error from `tree-sitter`, we can’t construct an `Expression`. Under a-la-carte syntax, all the children of a node were polymorphic, so we were able just to stick an `Error` node in there. But our types are stronger now, so we can’t get away with just stuffing `Error`s everywhere. Instead, what we’re gonna generate is this:

```
data Plus f a = Plus { ann :: a, left :: f (Expression f a), right :: f (Expression f a) }
```

In this new precise world, `Plus` has kind `(* -> *) -> * -> *` where the first parameter `f` is a “shape” functor that dictates what possible shapes the children could have. If we start substituting values for `f`, we get different types of trees.  For example,`Plus Identity` means that `left :: Identity (Expression Identity a)`, `right :: Identity (Expression Identity a)`. 

Since `Identity a` is isomorphic to `a`, that means `Plus Identity` is isomorphic to our old definition of `Plus`.

But what happens when we start getting some more interesting functors? Let’s consider `Plus (Either String)`. Substituting in `Either String` for `f`, that leaves us with children that look like `left :: Either String (Expression (Either String) a)`, `right :: Either String (Expression (Either String) a)` and _that_ lets us handle parse errors. The left field is `Right` if the parse succeeded and `Left` if it failed. This means we can slurp a partial or erroneous tree via Unmarshal.

This PR also has a few additional changes that are:
- Cosmetic! 💄 Better readability, reducing repetition and using more expressive functions. 
- Standardizing: use the TH function convention instead of data constructors (ie., `conT` instead of `ConT`); introducing constructor names in signatures that otherwise had ambiguous types (ex., `DatatypeName` instead of `String`).

### CODE REVIEW

File: semantic-ast/src/AST/GenerateSyntax.hs [hunk 1]
```diff
@@ -62,76 +66,80 @@ getAllSymbols language = do
...
+
 -- Auto-generate Haskell datatypes for sums, products and leaf types
 syntaxDatatype :: Ptr TS.Language -> [(String, Named)] -> Datatype -> Q [Dec]
 syntaxDatatype language allSymbols datatype = skipDefined $ do
-  typeParameterName <- newName "a"
+  shapeFunctorKind <- [t| * -> * |]
+  let traversalInstances = mappend <$> makeStandaloneDerivings (conT name) <*> makeTraversalInstances (conT name)
+      glue a b c = a : b <> c
+      name = mkName nameStr
+      generatedDatatype cons = dataD (cxt []) name [kindedTV shapeParameterName shapeFunctorKind, plainTV annParameterName] Nothing cons [deriveStockClause, deriveAnyClassClause]
```
Comment (aymannadeem) [inline semantic-ast/src/AST/GenerateSyntax.hs:80]: We generate `f :: * -> *` parameters for the record types, in TH, and add them to the generated datatype.

File: semantic-ast/src/AST/GenerateSyntax.hs [hunk 2]
```diff
@@ -62,76 +66,80 @@ getAllSymbols language = do
...
-    ProductType (DatatypeName datatypeName) named children fields -> do
-      con <- ctorForProductType datatypeName typeParameterName children fields
-      symbolMatchingInstance <- symbolMatchingInstance allSymbols name named datatypeName
-      traversalInstances <- makeTraversalInstances (conT name)
-      pure
-        (  generatedDatatype name [con] typeParameterName
-        :  symbolMatchingInstance
-        <> traversalInstances)
+    SumType (DatatypeName _) _ subtypes ->
+      let types' = fieldTypesToNestedSum shapeParameterName subtypes
```
Comment (aymannadeem) [inline semantic-ast/src/AST/GenerateSyntax.hs:86]: We can't use the above defined `generatedDatatype` for the generation of sum types since sum types are newtypes (ex: `data JSONValue f a = JSONValue { getValue :: (Array f :+: String f :+: EtCetera f) a }`. This change was introduced [here](https://github.com/tree-sitter/haskell-tree-sitter/pull/214) to reduce the complexity of the original structure ([explained here](https://github.com/tree-sitter/haskell-tree-sitter/issues/210)). Due to this difference between sums and leaves, we can't resort to taking advantage of the factored out `generatedDatatype` function which produces a datatype, not a newtype.

File: semantic-ast/src/AST/GenerateSyntax.hs [hunk 3]
```diff
@@ -62,76 +66,80 @@ getAllSymbols language = do
...
-  instance Foldable $ty where
+  instance Traversable f => Foldable ($ty f) where
     foldMap = foldMapDefault1
-  instance Functor $ty where
+  instance Traversable f => Functor ($ty f) where
     fmap = fmapDefault1
-  instance Traversable $ty where
+  instance Traversable f => Traversable ($ty f) where
     traverse = traverseDefault1
   |]
```
Comment (aymannadeem) [inline semantic-ast/src/AST/GenerateSyntax.hs:130]: We fix the `Traversable`, `Foldable`, and `Functor` instances to account for the new kindedness.

File: semantic-ast/src/AST/GenerateSyntax.hs [hunk 4]
```diff
@@ -146,41 +154,51 @@ debugPrefix (name, Named)     = name
...
+  annotation = ("ann", varT annParameterName)
+  fieldList = map (fmap (toType)) fields
   childList = toList $ fmap toTypeChild children
+
+  inject t = varT shapeParameterName `appT` t
+
+  toType :: Field -> TypeQ
   toType (MkField required fieldTypes mult) =
-    let ftypes = fieldTypesToNestedSum fieldTypes `appT` varT typeParameterName
+    let ftypes = inject (fieldTypesToNestedSum shapeParameterName fieldTypes `appT` varT annParameterName)
```
Comment (aymannadeem) [inline semantic-ast/src/AST/GenerateSyntax.hs:168]: ensure that the children of each record types are wrapped in an `f` (except for the ann field)

File: semantic-typescript/src/Language/TypeScript/Tags.hs [hunk 5]
```diff
@@ -24,19 +30,19 @@ import Source.Source as Source
...
       Has (Writer Tags.Tags) sig m
     ) =>
-    t Loc ->
+    t Parse.Err Loc ->
     m ()
   default tags ::
     ( Has (Reader Source) sig m,
       Has (Writer Tags.Tags) sig m,
-      Traversable1 ToTags t
+      Traversable1 ToTags (t Parse.Err)
```
Comment (aymannadeem) [inline semantic-typescript/src/Language/TypeScript/Tags.hs:43]: I added `Parse.Err` to this constraint in response to a directive issued by a prior error message:

```
    • Expecting one more argument to ‘Traversable1 (ToTags t)’
      Expected a constraint,
        but ‘Traversable1 (ToTags t)’ has kind ‘(* -> *) -> Constraint’
    • In the type signature:
        tags :: (Has (Reader Source) sig m,
                 Has (Writer Tags.Tags) sig m,
                 Traversable1 (ToTags t)) =>
                t Parse.Err Loc -> m ()
      In the class declaration for ‘ToTags’
   |
43 |       Traversable1 (ToTags t)
   |       ^^^^^^^^^^^^^^^^^^^^^^^
```

However,  this addition only results in the following error:

```
Tags.hs:43:20-25: error:
    • Expected kind ‘(* -> *) -> Constraint’,
        but ‘ToTags’ has kind ‘((* -> *) -> * -> *) -> Constraint’
    • In the first argument of ‘Traversable1’, namely ‘ToTags’
      In the type signature:
        tags :: (Has (Reader Source) sig m,
                 Has (Writer Tags.Tags) sig m,
                 Traversable1 ToTags (t Parse.Err)) =>
                t Parse.Err Loc -> m ()
      In the class declaration for ‘ToTags’
   |
43 |       Traversable1 ToTags (t Parse.Err)
   | 
```

I hypothesized that the `ConstraintKinds` extension could be used to treat constraints as first-class citizens, but I think I'm conflating a few too many things about constraint kinds. @patrickt could use some 🍐-ing help here.

**Comment by aymannadeem** [comment]
Update: none of the kind adjustments happened due to the new shape being in conflict with `Traversable1`. However, this PR accomplishes adding an error functor `Err`.

File: .github/workflows/haskell.yml [hunk 6]
```diff
@@ -62,8 +62,8 @@ jobs:
         cabal v2-run   --project-file=cabal.project.ci semantic-java:test
         cabal v2-run   --project-file=cabal.project.ci semantic-json:test
         cabal v2-run   --project-file=cabal.project.ci semantic-python:test
-        cabal v2-run   --project-file=cabal.project.ci semantic-python:test:compiling
-        cabal v2-run   --project-file=cabal.project.ci semantic-python:test:graphing
+        # cabal v2-run   --project-file=cabal.project.ci semantic-python:test:compiling
+        # cabal v2-run   --project-file=cabal.project.ci semantic-python:test:graphing
```
Comment (patrickt) [inline .github/workflows/haskell.yml:66]: We should consult with @joshvera and @BekaValentine about how/when we want to merge the efforts in this branch with the scope-graphing ones. We could make this branch based off of the stack graph one, or vice versa; either way, I’m happy to help with the plug-and-chug addition of `Success` branches.

File: semantic-ast/src/AST/Parse.hs [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+-- data If f a = If { ann :: a, condition :: f (Expr f a), consequence :: f (Stmt f a), alternative :: f (Stmt f a) }
+-- @
+-- When the parse fails, the f will be substituted with Err
+-- TODO: this could work with AST.Element Prj given the kindedness was adjusted from (*) to (* -> *).
+data Err a = Fail String | Success a
+  deriving (Eq, Foldable, Functor, Generic, Generic1, Ord, Traversable)
+
+instance Applicative Err where
+    pure            = Success
+    Fail e <*> _    = Fail e
```
Comment (patrickt) [inline semantic-ast/src/AST/Parse.hs:27]: Might be worth noting that this instance prefers the error message on the LHS of the `<*>` operator. Alternatively, we could do

```
data Err a = Fail (NonEmpty String) | Success a
```

and say that the `Applicative` instance over `Fail` concatenates the reasons for failure.

File: semantic-php/src/Language/PHP/Tags.hs [hunk 8]
```diff
@@ -5,13 +5,15 @@
 {-# LANGUAGE OverloadedLists #-}
 {-# LANGUAGE TypeApplications #-}
 {-# LANGUAGE TypeOperators #-}
+{-# OPTIONS_GHC -Wno-incomplete-patterns #-}
```
Comment (patrickt) [inline semantic-php/src/Language/PHP/Tags.hs:8]: Better 🔥 this; I think you’ve fixed most of the patterns here, if not all of them.

File: semantic-python/semantic-python.cabal [hunk 9]
```diff
@@ -58,59 +58,60 @@ library
   exposed-modules:
     Language.Python
     Language.Python.AST
-    Language.Python.Core
+    -- Language.Python.Core
```
Comment (patrickt) [inline semantic-python/semantic-python.cabal:61]: Can we file a bug to reenable the Core stuff (or elide it entirely for the time being)?

File: semantic-python/src/Language/Python/Tags.hs [hunk 10]
```diff
@@ -97,46 +98,66 @@ instance ToTags Py.FunctionDefinition where
...
   src <- ask @Source
   Tags.yield (Tag name kind ty loc (Tags.firstLine src srcLineRange) docs)
 
-docComment :: Source -> (Py.CompoundStatement :+: Py.SimpleStatement) Loc -> Maybe Text
-docComment src (R1 (Py.SimpleStatement (Prj Py.ExpressionStatement {extraChildren = L1 (Prj (Py.Expression (Prj (Py.PrimaryExpression (Prj Py.String {ann}))))) :| _}))) = Just (toText (slice src (byteRange ann)))
+docComment :: Source -> Parse.Err ((Py.CompoundStatement :+: Py.SimpleStatement) Loc) -> Maybe Text
+docComment
+  src
+  ( Parse.Success
+      ( R1
```
Comment (patrickt) [inline semantic-python/src/Language/Python/Tags.hs:146]: I think this combination of `Success` and `R1` can be expressed with `EPrj` (and further on with that `L1` pattern match).

File: semantic-ruby/src/Language/Ruby/Tags.hs [hunk 11]
```diff
@@ -6,6 +6,7 @@
 {-# LANGUAGE TypeApplications #-}
 {-# LANGUAGE TypeOperators #-}
 {-# OPTIONS_GHC -Wno-unrecognised-pragmas #-}
+{-# OPTIONS_GHC -Wno-incomplete-uni-patterns #-}
```
Comment (patrickt) [inline semantic-ruby/src/Language/Ruby/Tags.hs:9]: Both of these are suspect— 🔥 both of em

File: semantic-ruby/src/Language/Ruby/Tags.hs [hunk 12]
```diff
@@ -261,14 +288,14 @@ instance ToTags Rb.MethodCall where
...
-        Prj Rb.Constant {text, ann} -> yield text P.CALL ann
-        Prj Rb.Operator {text, ann} -> yield text P.CALL ann
+      EPrj (Rb.Variable (Prj Rb.Identifier {text, ann})) -> yield text P.CALL ann
+      EPrj (Rb.Variable (Prj Rb.Constant {text, ann})) -> yield text P.CALL ann -- TODO: Should yield Constant
+      EPrj Rb.ScopeResolution {name = EPrj Rb.Identifier {text, ann}} -> yield text P.CALL ann
+      EPrj Rb.ScopeResolution {name = EPrj Rb.Constant {text, ann}} -> yield text P.CALL ann -- TODO: Should yield Constant
+      EPrj Rb.Call {method} -> case method of
+        EPrj Rb.Identifier {text, ann} -> yield text P.CALL ann
+        EPrj Rb.Constant {text, ann} -> yield text P.CALL ann
+        EPrj Rb.Operator {text, ann} -> yield text P.CALL ann
```
Comment (patrickt) [inline semantic-ruby/src/Language/Ruby/Tags.hs:298]: I love how little this code actually changed! 💯

File: semantic-ruby/src/Language/Ruby/Tags.hs (hunk 11 above)
Comment (aymannadeem) [reply semantic-ruby/src/Language/Ruby/Tags.hs:9]: cool, `{-# HLINT ignore "Reduce duplication" #-}` is ok?

File: semantic-python/semantic-python.cabal (hunk 9 above)
Comment (aymannadeem) [reply semantic-python/semantic-python.cabal:61]: https://github.com/github/semantic/issues/569

File: .github/workflows/haskell.yml (hunk 6 above)
Comment (joshvera) [reply .github/workflows/haskell.yml:66]: Happy to get this in and deal with the merge on the stack graph branch later.
